common/debug-interface/src/libra_trace.rs

```diff
@@ -296,3 +333,60 @@ fn abbreviate_crate(name: &str) -> &str {
         _ => name,
     }
 }
+
+// This is exact copy of similar function in log crate
+/// Sets libra trace config
+pub fn set_libra_trace(config: &HashMap<String, String>) -> Result<(), ()> {
+    unsafe {
+        match LIBRA_TRACE_STATE.compare_and_swap(UNINITIALIZED, INITIALIZING, Ordering::SeqCst) {
+            UNINITIALIZED => {
+                SAMPLING_CONFIG = Some(parse_sampling_config(config));
+                LIBRA_TRACE_STATE.store(INITIALIZED, Ordering::SeqCst);
+                Ok(())
+            }
+            INITIALIZING => {
+                while LIBRA_TRACE_STATE.load(Ordering::SeqCst) == INITIALIZING {}
+                Err(())
+            }
+            _ => Err(()),
+        }
+    }
+}
+
+fn parse_sampling_config(config: &HashMap<String, String>) -> Sampling {
+    Sampling(
+        config
+            .iter()
+            .map(|(k, v)| {
+                let k: &'static str = Box::leak(k.clone().into_boxed_str());
+                let v = v.split('/').collect::<Vec<&str>>();
+                let v = CategorySampling {
+                    nominator: v[0].parse::<u64>().unwrap(),
+                    denominator: v[1].parse::<u64>().unwrap(),
+                };
+                (k, v)
+            })
+            .collect::<HashMap<&'static str, CategorySampling>>(),
+    )
+}
+
+/// Checks if libra trace is enabled
+pub fn libra_trace_set() -> bool {
+    LIBRA_TRACE_STATE.load(Ordering::SeqCst) == INITIALIZED
+}
+
+pub fn is_selected(node: (&str, &Vec<u8>)) -> bool {
+    unsafe {
+        match &SAMPLING_CONFIG {
+            Some(Sampling(sampling)) => {
+                // Take the leading 6 hex of account address or block hashvalue to hash
```

Comment:
I don't think we can take first few letters of value and parse them as hex - that's not very universal and seem tailored specifically for txn/blocks.

What we can do instead is this: instead of `parse_node` have macro `node_sampling_data -> (&'static str, u64)`, where `&'static str` will be same as what you currently return (first element) and `u64` will be proper hash of remaining elements

This 'proper hash' can be calculated by initializing `std::collections::hash_map::DefaultHasher` and then calling `.hash(&mut hasher)` on each remaining element of the node.

You can see `format_struct_args` in logger create for an example of how you can recursively iterate through `$rest` in a macro. (example is logger create much more scary, in your case it should be much shorter).

One of the problems in current approach you are taking(aside from it not being universal) is that you calculate transaction sampling based on block id (first element in tuple). That's pretty bad as it would generate big spike of logging for each block that won the lottery